tools/tiled_ensemble/predictions/prediction_data.py

```diff
@@ -0,0 +1,286 @@
+"""Classes used to store ensemble predictions."""
+
+# Copyright (C) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+import copy
+from abc import ABC
+from pathlib import Path
+from typing import Dict, List, Tuple
+
+import torch
+import torch.nn.functional as F
+from torch import Tensor
+
+
+class EnsemblePredictions(ABC):
+    """Abstract class used as template for different ways of storing ensemble predictions."""
+
+    def __init__(self) -> None:
+        self.num_batches = 0
+
+    def add_tile_prediction(self, tile_index: Tuple[int, int], tile_prediction: List[Dict[str, Tensor | List]]) -> None:
+        """
+        Add tile prediction data at specified tile index.
+
+        Args:
+            tile_index: Index of tile that we are adding in form (row, column).
+            tile_prediction: List of batches containing all predicted data for current tile.
+        """
+        raise NotImplementedError
+
+    def get_batch_tiles(self, batch_index: int) -> Dict[Tuple[int, int], Dict]:
+        """
+        Get all tiles of current batch.
+
+        Args:
+            batch_index: Index of current batch of tiles to be returned.
+
+        Returns:
+            Dictionary mapping tile index to predicted data, for provided batch index.
+        """
+        raise NotImplementedError
+
+
+class BasicEnsemblePredictions(EnsemblePredictions):
+    """Basic implementation of EnsemblePredictionData that keeps all predictions in memory as they are.
+
+    Examples:
+        >>> from pytorch_lightning import Trainer
+        >>> data = BasicEnsemblePredictions()
+        >>> trainer = Trainer(...)
+        >>>
+        >>> # predictions for all batches for tile location can be added
+        >>> curr_predictions = trainer.predict()
+        >>> # curr_predictions is list of batches predicted for index
+        >>> data.add_tile_prediction((0, 0), curr_predictions)
+
+        >>> # all tile location predictions can then be obtained for each batch
+        >>> batch_data = data.get_batch_tiles(0)
+        >>> # batch_data is dictionary where keys are all stored tile indices, and values are batch
+        >>> # of predictions at tile location for current index
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.all_data: Dict[Tuple[int, int], List] = {}
+
+    def add_tile_prediction(self, tile_index: Tuple[int, int], tile_prediction: List[Dict[str, Tensor | List]]) -> None:
```

Comment:
```suggestion
    def add_tile_prediction(self, tile_index: tuple[int, int], tile_prediction: list[Dict[str, Tensor | list]]) -> None:
```

And for the others please

Reply:
Done